### PR TITLE
Improving tests

### PR DESCRIPTION
**Pull request by macartur**

### CODE REVIEW

**Comment by cemsbr** [comment]
Great, you got several bugs, including typos that are hard to notice!

File: pyof/foundation/base.py [hunk 1]
```diff
@@ -558,7 +560,13 @@ def update_header_length(self):
         When sending an OpenFlow message we need to inform the message length
         on the header. This is mandatory.
         """
-        self.header.length = self.get_size()
+        if self.header is not None:
```
Comment (diraol) [inline pyof/foundation/base.py]: @macartur I think that here is better to use a try except condition, since we expect all messages  to have a header attribute.

On the new __init_subclass__ I'm proposing this will be more easily handled, since the header existence will be checked during the class creation and, therefore, will probably exist when the update_header_length method get called.

But I am not against doing a check over here. Can you rewrite this part on the commit that added this `if`?

You can force push the change, no problem with that.

File: pyof/foundation/base.py [hunk 2]
```diff
@@ -558,7 +560,13 @@ def update_header_length(self):
...
+            self.header.length = self.get_size()
+
+    def __setattr__(self, name, value):
+        """Update the header lenght when a attribute is updated."""
+        super().__setattr__(name, value)
+        self.update_header_length()
```
Comment (diraol) [inline pyof/foundation/base.py]: Hum.....
I liked the idea, but I'm not sure if this is going to work, and, more than that, I think that this will introduce a false idea that the length (header.length) will be always updated.

Think with me.

Consider the message (class) `A`, that have an attribute `b`, which is an instance of the struct (class) `B`.
If you change an attribute of `b`, the `A.__setattr__` will never be called, and then we can have a disparity between the header.length value and the real length of the message.

Using our classes as examples:

```python
from pyof.v0x01.common.actions import ActionOutput
from pyof.v0x01.controller2switch.flow_mod import FlowMod

mod = FlowMod()
action = ActionOutput()
mod.actions.append(action)
```
On this case, the final size of the message was altered, but the `header.length` will not be updated consistently.

So, I think it is better for us not to have this functionality if it can introduce a not always valid behavior.

File: pyof/v0x01/controller2switch/stats_reply.py [hunk 3]
```diff
@@ -31,6 +35,24 @@ def __init__(self, xid=None, body_type=None, flags=None, body=b''):
         self.flags = flags
         self.body = body
 
+    def pack(self):
```
Comment (diraol) [inline pyof/v0x01/controller2switch/stats_reply.py]: Our pack methods signature contains a `value` argument. Please, fix this to keep the method signature consistent.

File: pyof/v0x01/controller2switch/stats_reply.py [hunk 4]
```diff
@@ -31,6 +35,24 @@ def __init__(self, xid=None, body_type=None, flags=None, body=b''):
...
+        buff = self.body
+
+        if self.body and hasattr(self.body, 'pack'):
+            self.body = BinaryData(buff.pack())
+        stats_reply_packed = super().pack()
+        self.body = buff
+
+        return stats_reply_packed
+
     def unpack(self, buff):
```
Comment (diraol) [inline pyof/v0x01/controller2switch/stats_reply.py]: This unpack method is missing the 'offset' parameter.

File: pyof/v0x01/controller2switch/stats_reply.py [hunk 5]
```diff
@@ -47,17 +69,14 @@ def unpack(self, buff):
                 header.
         """
         super().unpack(buff)
+        self._unpack_body()
+
+    def _unpack_body(self):
+        """Unpack `body` replace it by the result."""
+        if self.body_type.value == 0:
```
Comment (diraol) [inline pyof/v0x01/controller2switch/stats_reply.py]: If we need to check for `OFPST_DESC` here, because it have a different behavior, then check for it, and not for a 'position' (`0`). If someone change the `_type` order at any time, then we would have a problem here.

File: pyof/v0x01/controller2switch/stats_request.py [hunk 6]
```diff
@@ -25,6 +26,9 @@ class StatsRequest(GenericMessage):
     flags = UBInt16()
     body = BinaryData()
 
+    _types = [None, FlowStatsRequest, AggregateStatsRequest,
```
Comment (diraol) [inline pyof/v0x01/controller2switch/stats_request.py]: Where does this `None`s came from? Why they are on the list?

File: pyof/v0x01/controller2switch/stats_request.py [hunk 7]
```diff
@@ -44,29 +48,20 @@ def pack(self):
         Make `body` a binary pack before packing this object. Then, restore
```
Comment (diraol) [inline pyof/v0x01/controller2switch/stats_request.py]: Please, remember that the `pack` method must have a `value` attribute on the signature.

File: pyof/v0x01/controller2switch/stats_request.py [hunk 8]
```diff
@@ -44,29 +48,20 @@ def pack(self):
...
-           self.body_type == StatsTypes.OFPST_FLOW or \
-           self.body_type == StatsTypes.OFPST_AGGREGATE:
-            backup = self.body
-            self.body = self.body.pack()
-            pack = super().pack()
-            self.body = backup
-            return pack
-        else:
+        if not self._types[self.body_type.value]:
             return super().pack()
```
Comment (diraol) [inline pyof/v0x01/controller2switch/stats_request.py]: (and the `value` attribute may be passed here  also)

File: pyof/v0x01/controller2switch/stats_request.py [hunk 9]
```diff
@@ -44,29 +48,20 @@ def pack(self):
...
 
+        backup = self.body
+        if hasattr(self.body, 'pack'):
+            self.body = self.body.pack()
+        pack = super().pack()
```
Comment (diraol) [inline pyof/v0x01/controller2switch/stats_request.py]: If you are going to edit this commit, then if you can use a better name for the `pack` variable it would be nice! (I know it was already there). may be `packed` or `data` or something that does not cause a confusion between the method and the variable. ;)

File: pyof/v0x01/controller2switch/stats_request.py [hunk 10]
```diff
@@ -44,29 +48,20 @@ def pack(self):
...
+        self.body = backup
+
+        return pack
+
     def unpack(self, buff):
```
Comment (diraol) [inline pyof/v0x01/controller2switch/stats_request.py]: Missing `offset` argument.

File: tests/test_struct.py [hunk 11]
```diff
@@ -107,7 +107,8 @@ def get_raw_object(cls):
             A new object using class and parameters priviously set through
                 :meth:`set_raw_dump_object`.
         """
-        return cls._new_raw_object()
+        obj = cls._new_raw_object()
+        return obj
```
Comment (diraol) [inline tests/test_struct.py]: Just to understand, why the change here?

File: tests/v0x01/test_common/test_header.py [hunk 12]
```diff
@@ -34,11 +34,11 @@ def test_pack(self):
     def test_unpack(self):
         """[Common/Header] - unpacking Hello."""
         filename = os.path.join(os.path.dirname(os.path.realpath('__file__')),
-                                'raw/v0x01/ofpt_hello.dat')
+                                'raw/v0x01/ofpt_header.dat')
```
Comment (diraol) [inline tests/v0x01/test_common/test_header.py]: @macartur, I think that there is not `ofpt_header` type for a `ofpt_header.dat` file. We were using the `ofpt_hello.dat` because the `Hello` Message only have the `Header` attribute.

File: tests/v0x01/test_common/test_header.py [hunk 13]
```diff
@@ -34,11 +34,11 @@ def test_pack(self):
...
         f = open(filename, 'rb')
         self.message.unpack(f.read(8))
 
-        self.assertEqual(self.message.length, 8)
+        self.assertEqual(self.message.length, 0)
```
Comment (diraol) [inline tests/v0x01/test_common/test_header.py]: The header cannot have length 0. This does not makes sense...

**Comment by beraldoleal** [comment]
@diraol is this ready to accept ?
